[file_selector] Add macOS support

stuartmorgan opened the pull request:
Brings file_selector_macos into flutter/plugins from FDE, with the following changes:
- Refactored slightly to allow for unit tests of almost all of the native code
- Added native unit test coverage
- Translated to Swift, to follow repo conventions for macOS plugins
- Added an in-package example (almost an exact duplicate of the app-facing version, but written against the platform interface, as is our current practice)
- Moved to an in-package method channel. As part of that, moved the flattening of type groups from Swift to Dart.

Does not currently include native UI tests to allow for end-to-end testing (since Flutter integration tests can't be used). They should be added later (they are currently blocked on https://github.com/flutter/flutter/issues/90673), but the unit tests give substantial coverage, making it substantially better to move the plugin now to get those tests running.

macOS portion of https://github.com/flutter/flutter/issues/70221

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

stuartmorgan commented:
@cbracken There are two options for review here:
1. As a full review
2. As diffs from what's in FDE, using the incremental commits

2 is complicated somewhat by the fact that after I had the unit tests in place I translated the plugin to Swift, but that's its own commit so it can be done as a side-by-side.

Either way:
- You can pretty much ignore `example/lib/` unless you have a strong desire to review it, because it's just a clone of `file_selector/file_selector/example/lib/` with very minor changes to use a different interface (something we're trying to figure out a good solution for in federated plugins in general, but don't currently have).
- All of `example/macos/` is just boilerplate `flutter create`, other than the unit test file (and the addition of a unit test target)

ditman commented:
(Maybe some day :P)

Hixie commented:
What's the status of this PR?

stuartmorgan commented:
> What's the status of this PR?

Waiting for @cbracken to have time to review it.

cbracken commented on packages/file_selector/file_selector_macos/example/lib/get_directory_page.dart:
```diff
@@ -0,0 +1,76 @@
+import 'package:file_selector_platform_interface/file_selector_platform_interface.dart';
+import 'package:flutter/material.dart';
+
+/// Screen that shows an example of getDirectoryPath
```
See comments on the [Windows patch](https://github.com/flutter/plugins/pull/4424#pullrequestreview-862798634)

stuartmorgan commented:
@cbracken This is ready for review:
- I updated the example to match the Windows changes
- I added the in-package method channel
- I reworked the native code and tests to put a stub-able protocol between the plugin implementation and the `FlutterPluginRegistrar` because it turns out we can't always support both `master` and `stable` with direct stubs. Master added a new method that uses a new type recently, and a stub would have to implement that method... but can't because the type of one of its arguments doesn't exist on stable. (And more generally, we'd be broken any time a new method were added to `FlutterPluginRegistrar`, and would have to update the tests, which isn't great.)

cbracken commented on packages/file_selector/file_selector_macos/example/lib/save_text_page.dart:
```diff
@@ -0,0 +1,78 @@
+  Future<void> _saveFile() async {
+    final String fileName = _nameController.text;
+    final String? path = await FileSelectorPlatform.instance.getSavePath(
+      // Operation was canceled by the user.
```
Move comment a few lines down.

stuartmorgan replied:
🤦🏻

cbracken reviewed (approved):
lgtm just two minuscule nits.

stuartmorgan commented on packages/file_selector/file_selector_macos/example/lib/get_directory_page.dart:
```diff
@@ -0,0 +1,77 @@
+import 'package:flutter/material.dart';
+
+/// Screen that allows the user to select a directory using `getDirectoryPath`,
+///  then displays the selected directory in a dialog.
```
Done.
